File: app/src/main/java/org/isoron/uhabits/ui/habits/show/views/HistoryCard.java
/*
 * Copyright (C) 2016 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.ui.habits.show.views;

import android.content.*;
import android.util.*;
import android.widget.*;

import org.isoron.uhabits.*;
import org.isoron.uhabits.models.*;
import org.isoron.uhabits.tasks.*;
import org.isoron.uhabits.ui.common.views.*;
import org.isoron.uhabits.utils.*;

import butterknife.*;

public class HistoryCard extends HabitCard
{
    @BindView(R.id.historyChart)
    HistoryChart chart;

    @BindView(R.id.title)
    TextView title;

    public HistoryCard(Context context)
    {
        super(context);
        init();
    }

    public HistoryCard(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        init();
    }

    @OnClick(R.id.edit)
    public void onClickEditButton()
    {
        Log.d("HistoryCard", "onClickEditButton");

//        HistoryEditorDialog frag = new HistoryEditorDialog();
//        frag.setHabit(habit);
//        frag.show(getContext().getFragmentManager(), "historyEditor");
    }

    @Override
    protected void refreshData()
    {
        Habit habit = getHabit();

        new BaseTask()
        {
            @Override
            protected void doInBackground()
            {
                int checkmarks[] = habit.getCheckmarks().getAllValues();
                chart.setCheckmarks(checkmarks);
            }

            @Override
            protected void onPreExecute()
            {
                super.onPreExecute();
                int color = ColorUtils.getColor(getContext(), habit.getColor());
                title.setTextColor(color);
                chart.setColor(color);
            }
        }.execute();
    }

    private void init()
    {
        inflate(getContext(), R.layout.show_habit_history, this);
        ButterKnife.bind(this);

        if (isInEditMode()) initEditMode();
    }

    private void initEditMode()
    {
        int color = ColorUtils.getAndroidTestColor(1);
        title.setTextColor(color);
        chart.setColor(color);
        chart.populateWithRandomData();
    }
}

File: android/uhabits-core/src/main/java/org/isoron/uhabits/core/ui/screens/habits/show/views/HistoryCard.kt
/*
 * Copyright (C) 2016-2020 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.core.ui.screens.habits.show.views

import org.isoron.platform.time.LocalDate
import org.isoron.uhabits.core.models.Entry
import org.isoron.uhabits.core.models.Entry.Companion.SKIP
import org.isoron.uhabits.core.models.Entry.Companion.YES_AUTO
import org.isoron.uhabits.core.models.Entry.Companion.YES_MANUAL
import org.isoron.uhabits.core.models.Habit
import org.isoron.uhabits.core.models.PaletteColor
import org.isoron.uhabits.core.ui.views.HistoryChart
import org.isoron.uhabits.core.ui.views.Theme
import org.isoron.uhabits.core.utils.DateUtils
import kotlin.math.max

data class HistoryCardViewModel(
    val color: PaletteColor,
    val firstWeekday: Int,
    val series: List<HistoryChart.Square>,
    val theme: Theme,
    val today: LocalDate,
)

class HistoryCardPresenter {
    fun present(
        habit: Habit,
        firstWeekday: Int,
        isSkipEnabled: Boolean,
        theme: Theme,
    ): HistoryCardViewModel {
        val today = DateUtils.getTodayWithOffset()
        val oldest = habit.computedEntries.getKnown().lastOrNull()?.timestamp ?: today
        val entries = habit.computedEntries.getByInterval(oldest, today)
        val series = if (habit.isNumerical) {
            entries.map {
                Entry(it.timestamp, max(0, it.value))
            }.map {
                when (it.value) {
                    0 -> HistoryChart.Square.OFF
                    else -> HistoryChart.Square.ON
                }
            }
        } else {
            entries.map {
                when (it.value) {
                    YES_MANUAL -> HistoryChart.Square.ON
                    YES_AUTO -> HistoryChart.Square.DIMMED
                    SKIP -> HistoryChart.Square.HATCHED
                    else -> HistoryChart.Square.OFF
                }
            }
        }

        return HistoryCardViewModel(
            color = habit.color,
            firstWeekday = firstWeekday,
            today = today.toLocalDate(),
            theme = theme,
            series = series,
        )
    }
}
